rust/src/io/commit.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trait for commit implementations.
+//!
+//! In Lance, a transaction is committed by writing the next manifest file.
+//! However, care should be taken to ensure that the manifest file is written
+//! only once, even if there are concurrent writers. Different stores have
+//! different abilities to handle concurrent writes, so a trait is provided
+//! to allow for different implementations.
+//!
+//! The trait [CommitHandler] can be implemented to provide different commit
+//! strategies. The default implementation for most object stores is
+//! [RenameCommitHandler], which writes the manifest to a temporary path, then
+//! renames the temporary path to the final path if no object already exists
+//! at the final path. This is an atomic operation in most object stores, but
+//! not in AWS S3. So for AWS S3, the default commit handler is
+//! [UnsafeCommitHandler], which writes the manifest to the final path without
+//! any checks.
+
+use std::{fmt::Debug, sync::atomic::AtomicBool};
+
+use crate::{format::Index, format::Manifest};
+use futures::future::BoxFuture;
+use object_store::path::Path;
+use object_store::Error as ObjectStoreError;
+
+use super::ObjectStore;
+
+/// Function that writes the manifest to the object store.
+pub type ManifestWriter = for<'a> fn(
+    object_store: &'a ObjectStore,
+    manifest: &'a mut Manifest,
+    indices: Option<Vec<Index>>,
+    path: &'a Path,
+) -> BoxFuture<'a, crate::Result<()>>;
+
+/// Handle commits that prevent conflicting writes.
+///
+/// Commit implementations ensure that if there are multiple concurrent writers
+/// attempting to write the next version of a table, only one will win. In order
+/// to work, all writers must use the same commit handler type.
+#[async_trait::async_trait]
+pub trait CommitHandler: Debug + Send + Sync {
+    /// Commit a manifest to a path.
+    ///
+    /// This function should return an [CommitError::CommitConflict] if another
+    /// transaction has already been committed to the path.
+    async fn commit(
+        &self,
+        manifest: &mut Manifest,
+        indices: Option<Vec<Index>>,
+        path: &Path,
+        object_store: &ObjectStore,
+        manifest_writer: ManifestWriter,
+    ) -> Result<(), CommitError>;
+}
```

Comment:
@chebbyChefNEQ What do you think of this? Does this trait give us enough flexibility for LanceDB to integrate with the catalog?

Reply:
I think is good. For being more ergonomic tho, would it make sense to add something like
```rust
trait CommitLock {
  async fn lock(...)
}

impl<T: CommitLock> CommitHandler for T {
...
}
```

Reply:
I was trying to be general, since the default implementations don't use locks. But I suppose almost all imaginable user-provided ones will, so maybe that should be the public API.